### PR TITLE
base: lmp-el2go-auto-register: replace ssscli

### PR DESCRIPTION
**Pull request by ldts**
Replace ssscli with fio-se05x-cli.
We should notice a performance improvement

### CODE REVIEW

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 1]
```diff
@@ -100,31 +101,20 @@ class Pkcs11:
...
-            out = subprocess.check_output(["ssscli", "se05x", "readidlist"])
-            lines = out.decode().splitlines()
-            # lines are like: Key-Id: 0Xf0000003   BINARY .....
-            return [x.split(" ")[1] for x in lines if x]
+    def read_id_list(self) -> List[str]:
+    command = ["fio-se05x-cli", "--list-objects", "all"]
+    process = subprocess.Popen(command, stdout=None, stderr=subprocess.PIPE)
+    stdout,stderr = process.communicate()
+    lines = stderr.decode().splitlines()
+    pattern = r'Key-Id:\s+(0x[0-9a-fA-F]+)'
```
Comment (MrCry0) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:111]: if a key-id always has the same length, a quantifier `{N}` has better performance than `+`

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 2]
```diff
@@ -100,31 +101,20 @@ class Pkcs11:
...
+class FioCli:
     @classmethod
-    def read_id_list(cls) -> List[str]:
-        with cls.i2c_session():
-            out = subprocess.check_output(["ssscli", "se05x", "readidlist"])
-            lines = out.decode().splitlines()
-            # lines are like: Key-Id: 0Xf0000003   BINARY .....
-            return [x.split(" ")[1] for x in lines if x]
+    def read_id_list(self) -> List[str]:
+    command = ["fio-se05x-cli", "--list-objects", "all"]
```
Comment (MrCry0) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:107]: python style needs a shift for each nested control block

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 3]
```diff
@@ -100,31 +101,20 @@ class Pkcs11:
...
+    pattern = r'Key-Id:\s+(0x[0-9a-fA-F]+)'
+    return [re.search(pattern,x).group(1) for x in lines if x]
 
     @classmethod
-    def get_cert(cls, keyid: str, certfile: str):
-        with cls.i2c_session():
-            subprocess.check_call(["ssscli", "get", "cert", keyid, certfile])
-
+    def import_cert(self, oid: str, identifier: str, label: str):
+    args = ["fio-se05x-cli", "--token-label", "aktualizr", "--import-cert", oid, "--id", identifier, "--label", label]
```
Comment (MrCry0) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:116]: python style needs shift for each nested control block

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register (hunk 1 above)
Comment (ldts) [reply meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:111]: the key id doesnt always have the same length as a string (it is a string representing any 32bit hex value)
reduced matches? um, I did try the above code with different lengths and it does as expected..what problem do you see with it?

Comment (MrCry0) [reply meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register]: > the key id doesnt always have the same length as a string (it is a string representing any 32bit hex value) reduced matches? um, I did try the above code with different lengths and it does as expected..what problem do you see with it?

I remind sometimes ago Python re had problems with its NFA engine. (Like described here https://bytes.com/topic/python/answers/534624-how-get-longest-possible-match-pythons-module)

They may have already fixed it in newer versions. But I personally never trust 3rd-party re algos (not perl/libpcre ;-) and guard match groups on both sides. Like `\s+(0x[0-9a-fA-F]+)\b`

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 4]
```diff
@@ -100,31 +88,20 @@ class Pkcs11:
...
-            out = subprocess.check_output(["ssscli", "se05x", "readidlist"])
-            lines = out.decode().splitlines()
-            # lines are like: Key-Id: 0Xf0000003   BINARY .....
-            return [x.split(" ")[1] for x in lines if x]
+    def read_id_list(self) -> List[str]:
+        command = ["fio-se05x-cli", "--list-objects", "all"]
+        process = subprocess.Popen(command, stdout=None, stderr=subprocess.PIPE)
+        stdout,stderr = process.communicate()
+        lines = stderr.decode().splitlines()
+        pattern = r'Key-Id:\s+(0x[0-9a-fA-F]+)'
```
Comment (doanac) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:98]: you are ignoring stdout and parsing stderr instead? Generally if you are ignoring a value in python you'd do:
```
_, stderr = process.communicate()
```

Also - python has a tool called "black". You can run it on this to auto-format it in canonical python3 style

Comment (MrCry0) [reply meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register]: > you are ignoring stdout and parsing stderr instead? Generally if you are ignoring a value in python you'd do:
> 
> ```
> _, stderr = process.communicate()
> ```
> 
> Also - python has a tool called "black". You can run it on this to auto-format it in canonical python3 style

Wow, thanks @doanac Black is amazing! :)

**Review by MrCry0 — APPROVED** [review]

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 5]
```diff
@@ -149,10 +139,7 @@ class AwsKpHandler(NamedTuple):
             return
         log.info("Processing keys for aws-iot...")
         pkcs11.gen_kp("04", key_label)
-        with NamedTemporaryFile() as f:
-            SssCli.get_cert(self.crt_id, f.name)
-            pkcs11.write_object("cert", f.name, "05", crt_label)
-
+        FioCli.import_cert(crt_id, "05", crt_label)
```
Comment (mwasilew) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:141]: `crt_id` should be `self.crt_id`

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 6]
```diff
@@ -181,9 +168,7 @@ class AkliteKpHandler(NamedTuple):
         else:
             log.info("Processing keys for aktualizr-lite...")
             pkcs11.gen_kp("01", key_label)
-            with NamedTemporaryFile() as f:
-                SssCli.get_cert(self.crt_id, f.name)
-                pkcs11.write_object("cert", f.name, "03", crt_label)
+            FioCli.import_cert(crt_id, "03", crt_label)
```
Comment (mwasilew) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:171]: `crt_id` should be `self.crt_id`

**Comment by mwasilew** [comment]
This works for factory registration. I didn't test AWS case yet.

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register [hunk 7]
```diff
@@ -101,29 +89,34 @@ class Pkcs11:
...
 
     @classmethod
-    def get_cert(cls, keyid: str, certfile: str):
-        with cls.i2c_session():
-            subprocess.check_call(["ssscli", "get", "cert", keyid, certfile])
+    def import_cert(self, oid: str, identifier: str, label: str):
+        args = [
+            "fio-se05x-cli",
+            "--token-label",
+            "aktualizr",
```
Comment (ricardosalveti) [inline meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:111]: Wonder if the hardcoded token label will cause us issues in the future, but I guess we're using it as standard all across, so should be fine.

**Comment by ricardosalveti** [comment]
Please squash all your commits.

Please also update the recipe itself to drop the dependencies on ssscli and opensc, adding fio-se05x-cli instead.

File: meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register (hunk 7 above)
Comment (ldts) [reply meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register:111]: I could do more improvements - ie, fio-se05x-cli does know if the keys and certificates are already in the pkcs#11 database so we could skip one of the pkcs#11 calls in the handler..but we can do that in the next release without having to parse so much...

**Comment by ldts** [comment]
> Please squash all your commits.
> 
> Please also update the recipe itself to drop the dependencies on ssscli and opensc, adding fio-se05x-cli instead.

but we still need opensc for pkcs11-tool

**Comment by ldts** [comment]
just squashed the latest changes. @rsalveti ok to merge?

**Comment by ricardosalveti** [comment]
> but we still need opensc for pkcs11-tool to import the keys into the pkcs11 database. I'll drop ssscli and the plug-and-trust stuff though

True, it is still required.

**Pull request merged**